akka-actor/src/main/scala/akka/io/dns/DnsResourceRecords.scala

```diff
@@ -156,9 +156,20 @@ private[dns] object ResourceRecord {
     val data = it.clone().take(rdLength)
     it.drop(rdLength)
     (recType: @switch) match {
-      case 1  => ARecord.parseBody(name, ttl, rdLength, data)
-      case 5  => CNameRecord.parseBody(name, ttl, rdLength, data, msg)
-      case 28 => AAAARecord.parseBody(name, ttl, rdLength, data)
+      case 1 => ARecord.parseBody(name, ttl, rdLength, data)
+      case 5 => CNameRecord.parseBody(name, ttl, rdLength, data, msg)
+      case 28 => {
+        val (it1, it2) = data.duplicate
+        try {
+          AAAARecord.parseBody(name, ttl, rdLength, it1)
+        } catch {
+          case _: java.lang.ClassCastException => {
```

Comment:
Throwing and catching is pretty resource-intensive. Could we change `parseBody` to turn the address bytes into an `Inet6Address` avoiding casting?

Reply:
To sure that I understand the request?
Do you want to return AAAARecord with undefined IP inside?
Smth like AAAARecord(name, ttl, InetAddress.getByAddress(address)])


```
if (addr.length == 4) {
                return new Inet4Address(host, addr);
            }

            if (addr.length == 16) {
                byte[] newAddr = IPAddressUtil.convertFromIPv4MappedAddress(addr);
                if (newAddr != null) {
                    return new Inet4Address(host, newAddr);
                }

                return new Inet6Address(host, addr);
            }
```

Anyway, for AAAARecord, ip should be of type Inet6Address, not InetAddress

Reply:
> Do you want to return AAAARecord with undefined IP inside?

Indeed I was thinking of putting the IPv4-mapped IPv6 address into the `Inet6Address` - https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/net/Inet6Address.html has a section on IPv4-mapped address, but to be honest I'm not sure if I should read "These classes can take an IPv4-mapped address as input, both in byte array and text representation. However, it will be converted into an IPv4 address." as meaning that we *can* or *shouldn't* put an IPv4-mapped IPv6 address into the `Inet6Address`. Probably the latter?

I would also be OK with changing the `Inet6Address` field of `AAAARecord` into an `InetAddress`, but since this is a case class that'd probably need some extra code to keep it binary compatible.

Reply:
I'll dig dipper, and return when I have something.